File: chapter06/src/main/java/com/example/chapter06/view/DiamondBackgroundTextView.java
package com.example.chapter06.view;

import android.content.Context;
import android.content.res.TypedArray;
import android.graphics.Canvas;
import android.graphics.Paint;
import android.graphics.Path;
import android.graphics.Rect;
import android.text.TextUtils;
import android.util.AttributeSet;
import android.view.View;

import androidx.annotation.Nullable;

import com.example.chapter06.R;

/**
 * @author wangzhichao
 * @since 2020/01/09
 */
public class DiamondBackgroundTextView extends View {

    private char[] arr;
    private Rect bounds = new Rect();
    private Rect backgroundRect = new Rect();
    private Path path = new Path();
    private Paint textPaint = new Paint(Paint.ANTI_ALIAS_FLAG);
    private Paint backgroundPaint = new Paint(Paint.ANTI_ALIAS_FLAG);
    private static final float OFFSET_PERCENT = 0.2f;
    private int baseLineX;
    private int baseLineY;

    public DiamondBackgroundTextView(Context context) {
        this(context, null);
    }

    public DiamondBackgroundTextView(Context context, @Nullable AttributeSet attrs) {
        this(context, attrs, 0);
    }

    public DiamondBackgroundTextView(Context context, @Nullable AttributeSet attrs, int defStyleAttr) {
        super(context, attrs, defStyleAttr);
        init(context, attrs);
    }

    private void init(Context context, @Nullable AttributeSet attrs ) {
        TypedArray ta = context.obtainStyledAttributes(attrs, R.styleable.DiamondBackgroundTextView);
        String text = ta.getString(R.styleable.DiamondBackgroundTextView_dbv_text);
        int backgroundColor = ta.getColor(R.styleable.DiamondBackgroundTextView_dbv_background_color, 0x80E5C8C8);
        int textColor = ta.getColor(R.styleable.DiamondBackgroundTextView_dbv_text_color, 0xFF7F101D);
        int textSize = ta.getDimensionPixelSize(R.styleable.DiamondBackgroundTextView_dbv_text_size, 100);
        int padding = ta.getDimensionPixelSize(R.styleable.DiamondBackgroundTextView_dbv_padding, 10);
        ta.recycle();
        textPaint.setColor(textColor);
        textPaint.setTextSize(textSize);
        textPaint.setTextAlign(Paint.Align.CENTER);
        backgroundPaint.setColor(backgroundColor);
        backgroundPaint.setStyle(Paint.Style.FILL);
        if (TextUtils.isEmpty(text)) {
            throw new IllegalArgumentException("no text");
        }
        arr = text.toCharArray();
        initBackgroundRect(padding);
        initBaseline();
        initDiamondPath();
    }

    @Override
    protected void onMeasure(int widthMeasureSpec, int heightMeasureSpec) {
        super.onMeasure(widthMeasureSpec, heightMeasureSpec);
        int width = measureWidth(widthMeasureSpec);
        int height = measureHeight(heightMeasureSpec);

        setMeasuredDimension(width, height);
    }

    /**
     * 决定 view 的高度
     *
     * @param heightMeasureSpec
     * @return
     */
    private int measureHeight(int heightMeasureSpec) {
        int heightMode = MeasureSpec.getMode(heightMeasureSpec);
        int heightSize = MeasureSpec.getSize(heightMeasureSpec);
        int result;
        if (heightMode == MeasureSpec.EXACTLY) {
            result = heightSize;
        } else {
            result = backgroundRect.height();
            if (heightMode == MeasureSpec.AT_MOST) {
                // 要求的高度不能超过父类给的高度
                result = Math.min(result, heightSize);
            }
        }
        return result;
    }

    /**
     * 决定 view 的宽度
     *
     * @param widthMeasureSpec
     * @return
     */
    private int measureWidth(int widthMeasureSpec) {
        int widthMode = MeasureSpec.getMode(widthMeasureSpec);
        int widthSize = MeasureSpec.getSize(widthMeasureSpec);
        int result;
        if (widthMode == MeasureSpec.EXACTLY) {
            result = widthSize;
        } else {
            float v = (arr.length - 1) * (1 - OFFSET_PERCENT) + 1;
            result = (int) (v * backgroundRect.width());
            if (widthMode == MeasureSpec.AT_MOST) {
                result = Math.min(result, widthSize);
            }
        }
        return result;
    }

    @Override
    protected void onDraw(Canvas canvas) {
        super.onDraw(canvas);
//        canvas.drawColor(Color.GREEN);
        for (int i = 0; i < arr.length; i++) {
            canvas.save();
            canvas.translate(backgroundRect.width() * (1 - OFFSET_PERCENT) * i, 0);
            canvas.drawPath(path, backgroundPaint);
            canvas.drawText(String.valueOf(arr[i]), baseLineX, baseLineY, textPaint);
            canvas.restore();
        }
    }

    private void initDiamondPath() {
        path.reset();
        path.moveTo(0, backgroundRect.centerY());
        path.lineTo(backgroundRect.centerX(), 0);
        path.lineTo(backgroundRect.width(), backgroundRect.centerY());
        path.lineTo(backgroundRect.centerX(), backgroundRect.height());
        path.close();
    }

    private void initBackgroundRect(int padding) {
        int side = getSide();
        backgroundRect.left = 0;
        backgroundRect.top = 0;
        backgroundRect.right = (padding * 2 + side) * 2;
        backgroundRect.bottom = (padding * 2 + side) * 2;
    }

    private void initBaseline() {
        Paint.FontMetricsInt fm = textPaint.getFontMetricsInt();
        baseLineX = backgroundRect.centerX();
        baseLineY = backgroundRect.centerY() - (fm.top + fm.bottom) / 2;
    }

    private int getSide() {
        int result = 0;
        for (int i = 0; i < arr.length; i++) {
            textPaint.getTextBounds(String.valueOf(arr[i]), 0, 1, bounds);
            int max = Math.max(bounds.width(), bounds.height());
            if (max > result) {
                result = max;
            }
        }
        return result;
    }
}

File: chapter06/src/main/java/com/example/chapter06/view/DiamondBackgroundTextView.kt
package com.example.chapter06.view

import android.content.Context
import android.graphics.*
import android.text.TextUtils
import android.util.AttributeSet
import android.view.View
import kotlin.jvm.JvmOverloads
import com.example.chapter06.R
import kotlin.math.max
import kotlin.math.min

/**
 * 中国结背景文字
 *
 * 没有处理超出一行的情况
 *
 * @author wangzhichao
 * @since 2020/01/09
 */
class DiamondBackgroundTextView @JvmOverloads constructor(
    context: Context,
    attrs: AttributeSet? = null,
    defStyleAttr: Int = 0
) : View(context, attrs, defStyleAttr) {
    /**
     * 要展示的字符数组
     */
    private val characterArray: CharArray
    private val bounds = Rect()
    private val backgroundRect = Rect()
    private val path = Path()
    private val textPaint = Paint(Paint.ANTI_ALIAS_FLAG)
    private val backgroundPaint = Paint(Paint.ANTI_ALIAS_FLAG)
    private var baseLineX = 0
    private var baseLineY = 0

    init {
        val ta = context.obtainStyledAttributes(attrs, R.styleable.DiamondBackgroundTextView)
        val text = ta.getString(R.styleable.DiamondBackgroundTextView_dbv_text)
        val backgroundColor =
            ta.getColor(R.styleable.DiamondBackgroundTextView_dbv_background_color, -0x7f1a3738)
        val textColor = ta.getColor(R.styleable.DiamondBackgroundTextView_dbv_text_color, -0x80efe3)
        val textSize =
            ta.getDimensionPixelSize(R.styleable.DiamondBackgroundTextView_dbv_text_size, 100)
        val padding =
            ta.getDimensionPixelSize(R.styleable.DiamondBackgroundTextView_dbv_padding, 10)
        ta.recycle()
        textPaint.color = textColor
        textPaint.textSize = textSize.toFloat()
        textPaint.textAlign = Paint.Align.CENTER
        backgroundPaint.color = backgroundColor
        backgroundPaint.style = Paint.Style.FILL
        require(!TextUtils.isEmpty(text)) { "no text" }
        characterArray = text!!.toCharArray()
        initBackgroundRect(padding)
        initBaseline()
        initDiamondPath()
    }

    override fun onMeasure(widthMeasureSpec: Int, heightMeasureSpec: Int) {
        super.onMeasure(widthMeasureSpec, heightMeasureSpec)
        val width = measureWidth(widthMeasureSpec)
        val height = measureHeight(heightMeasureSpec)
        setMeasuredDimension(width, height)
    }

    /**
     * 决定 view 的高度
     *
     * @param heightMeasureSpec
     * @return
     */
    private fun measureHeight(heightMeasureSpec: Int): Int {
        val heightMode = MeasureSpec.getMode(heightMeasureSpec)
        val heightSize = MeasureSpec.getSize(heightMeasureSpec)
        var result: Int
        if (heightMode == MeasureSpec.EXACTLY) {
            result = heightSize
        } else {
            result = backgroundRect.height()
            if (heightMode == MeasureSpec.AT_MOST) {
                // 要求的高度不能超过父类给的高度
                result = min(result, heightSize)
            }
        }
        return result
    }

    /**
     * 决定 view 的宽度
     *
     * @param widthMeasureSpec
     * @return
     */
    private fun measureWidth(widthMeasureSpec: Int): Int {
        val widthMode = MeasureSpec.getMode(widthMeasureSpec)
        val widthSize = MeasureSpec.getSize(widthMeasureSpec)
        var result: Int
        if (widthMode == MeasureSpec.EXACTLY) {
            result = widthSize
        } else {
            val v = (characterArray.size - 1) * (1 - OFFSET_PERCENT) + 1
            result = (v * backgroundRect.width()).toInt()
            if (widthMode == MeasureSpec.AT_MOST) {
                result = min(result, widthSize)
            }
        }
        return result
    }

    override fun onDraw(canvas: Canvas) {
        super.onDraw(canvas)
        for (i in characterArray.indices) {
            canvas.save()
            canvas.translate(backgroundRect.width() * (1 - OFFSET_PERCENT) * i, 0f)
            canvas.drawPath(path, backgroundPaint)
            canvas.drawText(characterArray[i].toString(), baseLineX.toFloat(), baseLineY.toFloat(), textPaint)
            canvas.restore()
        }
    }

    /**
     * 初始化菱形路径
     * 菱形是内嵌在正方形里面的。
     */
    private fun initDiamondPath() {
        path.reset()
        path.moveTo(0f, backgroundRect.centerY().toFloat())
        path.lineTo(backgroundRect.centerX().toFloat(), 0f)
        path.lineTo(backgroundRect.width().toFloat(), backgroundRect.centerY().toFloat())
        path.lineTo(backgroundRect.centerX().toFloat(), backgroundRect.height().toFloat())
        path.close()
    }

    /**
     * 初始化背景正方形
     */
    private fun initBackgroundRect(padding: Int) {
        backgroundRect.left = 0
        backgroundRect.top = 0
        backgroundRect.right = (padding * 2 + maxCharacterSide) * 2
        backgroundRect.bottom = (padding * 2 + maxCharacterSide) * 2
    }

    /**
     * 初始化绘制文字的原点
     */
    private fun initBaseline() {
        val fm = textPaint.fontMetricsInt
        baseLineX = backgroundRect.centerX()
        baseLineY = backgroundRect.centerY() - (fm.top + fm.bottom) / 2
    }

    /**
     * 获取背景正方形的边长
     */
    private val maxCharacterSide: Int
        get() {
            var result = 0
            for (character in characterArray) {
                // 包裹字符的最小矩形
                textPaint.getTextBounds(character.toString(), 0, 1, bounds)
                val max = max(bounds.width(), bounds.height())
                if (max > result) {
                    result = max
                }
            }
            return result
        }

    companion object {
        private const val OFFSET_PERCENT = 0.2f
    }

}